Switch from json to xml configuration

sebastianfeldmann opened the pull request.

sebastianfeldmann committed:
Switch from json to xml config

sebastianfeldmann committed:
Fix verify test without errors

theseer commented on composer-distributor.xsd:
```diff
@@ -0,0 +1,24 @@
+           targetNamespace="https://phar.io/composer-distributor" xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           xmlns:rep="https://phar.io/composer-distributor">
+
+  <xs:element name="composer-distributor" type="rep:distributorType"/>
```
Even though it's of course perfectly valid XML, I personally have a dislike for `-` within a tag name. Maybe we can (should?) shorten the root element name to `distributor`? That would also match the filename...

heiglandreas replied:
I've just seen that the w3c uses camelCased names (Or how are the camelCases called that don't have a first capital letter?). So perhaps `composerDistributor` would be an option? (see `targetNamespace`)

sebastianfeldmann replied:
Changed everything to `distributor` only :)

theseer commented on composer.json:
```diff
@@ -41,6 +41,8 @@
   "require": {
     "php": "^7.3 || ^8.0",
     "ext-json": "*",
```
Do we still need `ext/json`?

sebastianfeldmann replied:
Nope, good point I will remove it

theseer commented on src/Config/Loader.php:
```diff
@@ -2,33 +2,48 @@
     {
         if (!is_file($configFile)) {
-            throw new \RuntimeException('Config file is missing');
+            throw new RuntimeException('Config file is missing');
```
Do we actually want to consider throwing our own Exceptions instead?

heiglandreas replied:
```suggestion
            throw MissingConfigFile::fromFile($configFile);
```

sebastianfeldmann replied:
Own exceptions ✅

theseer commented on distributor.sample.xml:
```diff
@@ -0,0 +1,10 @@
+<composer-distributor xmlns="https://phar.io/composer-distributor"
+                      packageName="phar-io/phive"
+                      keyDirectory="keys">
+
```
We should probably add a schema location reference to this document:

```
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="https://phar.io/composer-distributor.xsd"`
```

Or something?

We should probably consider specifying a version identifier somehow...

heiglandreas replied:
* `xsi:schemaLocation="https://phar.io/composer-distributor.1.0.0.xsd`
* `xsi:schemaLocation="https://phar.io/2020/composer-distributor.xsd`
* `xsi:schemaLocation="https://phar.io/1.0.0/composer-distributor.xsd`
* `xsi:schemaLocation="https://phar.io/1/0/0/composer-distributor.xsd`

sebastianfeldmann replied:
what do you think about

    https://phar.io/data/composer-distributor/1.0/distributor.xsd

theseer replied:
So far, phar.io does not have any published schema urls, so we are relatively free to choose.

We also are sadly quite inconsistent already.

In terms of namespaces, a quick `ack` over all phar.io projects on my disc shows two patterns.

Phive:

- https://phar.io/auth
- https://phar.io/phive-auth
- https://phar.io/phive/installdb
- https://phar.io/phive
- https://phar.io/repository-list
- https://phar.io/repository

And our manifest:

- https://phar.io/xml/manifest/1.0


Sadly, I today think only the manifest format makes sense. :-(

In terms of schema location, it probably makes sense to derive that from the namespace.
So adopting your proposal:

`https://phar.io/xml/distributor/1.0/schema.xsd`

And I guess we just have to live with the other pattern in `phive` as changing that would be a major BC break with nothing to win :/

sebastianfeldmann replied:
`https://phar.io/xml/distributor/1.0/schema.xsd` it is

theseer commented on distributor.sample.xml:
```diff
@@ -0,0 +1,10 @@
+                      packageName="phar-io/phive"
+                      keyDirectory="keys">
+
+  <phar name="phive"
```
Do we actually need the `name` attribute?

heiglandreas replied:
That is what the binary will be renamed to. I'd rather have that explicitly set than implicitly somehow guestimated....

sebastianfeldmann replied:
Yes for the name of the binary since one package can hold multiple PHAR files

theseer commented on src/Config/Mapper.php:
```diff
@@ -4,93 +4,91 @@
+        \libxml_clear_errors();
+        \libxml_use_internal_errors($original);
+
+        if (count($errors) > 0) {
```
Given that `DOMDocument::schemaValidate()` returns a `bool` to signal success or failure, we should be able to exit early and only continue with error handling in case of `false`.

Or not?

sebastianfeldmann replied:
Done

theseer commented on src/Config/Mapper.php:
```diff
@@ -4,93 +4,91 @@
+        \libxml_use_internal_errors($original);
+
+        if (count($errors) > 0) {
+            throw new RuntimeException($this->formatErrorMessage($errors));
```
Again, use our own exception here?

sebastianfeldmann replied:
Done

theseer commented on src/Service/Verify.php:
```diff
@@ -50,18 +49,13 @@ public function fileWithSignature(SplFileInfo $file, SplFileInfo $signature) : b
         );
 
-        switch (true) {
-            case false === $result:
-            case $result[0]['summary'] !== 0:
```
Yoda, or not Yoda...? ;-)  I personally think Yoda should stay in Star Wars ;-p

theseer replied:
Disregard, the comment is about removed code ;)

theseer commented on src/Service/Verify.php:
```diff
@@ -50,18 +49,13 @@ public function fileWithSignature(SplFileInfo $file, SplFileInfo $signature) : b
-            default:
+        if ($result === false || $result[0]['summary'] !== 0) {
+            throw new RuntimeException(sprintf(
+                'Verification between "%s" and "%s" failed!',
+                $signature->getFilename(),
+                $file->getFilename()
```
Not sure if the error message is the best we can do.
"Verification between" doesn't sound like correct english to me..?

And: Own exception ;)

sebastianfeldmann replied:
Changed

theseer commented on tests/unit/Config/LoaderTest.php:
```diff
@@ -0,0 +1,29 @@
+{
+   public function testLoadConfigSuccess(): void
+   {
+        self::assertTrue(true);
```
w00t?

sebastianfeldmann replied:
Ups... missing test is added

theseer reviewed (changes requested):
Looks pretty good but of course got a few things I stumbled over..

heiglandreas commented on src/Config/Loader.php:
```diff
@@ -2,33 +2,48 @@
+        error_reporting($reporting);
+
+        if ($loaded === false || $message !== '') {
+            throw new RuntimeException(
```
```suggestion
            throw CouldNotLoad::fileWithMessages($filename, $message)
```

theseer replied:
TIL: \`\`\`suggestion is a thing ;)

sebastianfeldmann replied:
Custom exceptions all the way :)

heiglandreas reviewed (changes requested):
Own Exceptions are awesome ;-)

sebastianfeldmann committed:
Add custom exceptions

sebastianfeldmann committed:
Fix php-cs-fixer setup

sebastianfeldmann committed:
Add command shortcuts

sebastianfeldmann commented:
Is this merge-ready? If so we could start working on the mediator package and the setup wizard

heiglandreas commented on distributor.xsd:
```diff
@@ -0,0 +1,24 @@
+
+  <xs:complexType name="pharType">
+    <xs:attribute type="xs:string" name="name" use="required"/>
+    <xs:attribute type="xs:string"  name="file" use="required"/>
```
Shouldn't we use `anyUri` instead of `string` here?

sebastianfeldmann replied:
We cant't because of the %PLACEHOLDER% stuff :(
We would have to define a custom type.

heiglandreas replied:
Is that due to the `%`? If that is the issue I'd happily use something else as placeholder-marker. Like `::` or `#` or `$` or `@`

sebastianfeldmann replied:
Not sure because %PLACEHOLDER% is quite common, the next common thing would be {$PLACEHOLDER} I think (that would work in `anyURI`)

theseer replied:
What are we discussing here? Why should there be a %PLACEHOLDER%?

sebastianfeldmann replied:
URI can have placeholders in them so we can download the PHAR, the current implementation uses `%PLACEHOLDER%`
- http://github.com/releases/ %VERSION% /tool.phar
- http://github.com/releases/2.16.4/tool.phar

If we change the xsd definition from `xsd:string` to `xsd:anyURI` `%placeholder%` is not allowed because of the `%`
So we where thinking about changing the placeholder syntax or defining a custom type in the xsd.

theseer replied:
And we have to replace the version number at runtime, right?

sebastianfeldmann replied:
We don't have to, since you have to tag the mediating package as well we could force people to update the config file.

sebastianfeldmann replied:
But I would argue that allowing placeholders makes it much easier because you do not have to update the `distributor.xml` for every release. With placeholders you can just put the new `tag` on the same `commit`.

heiglandreas replied:
That was the initial idea. To be able to have as little work as possible with the plugin. Re-tagging should be enough. And the `%PLACEHOLDER%` syntax was choosen by me without further thinking! `{$PLACEHOLDER}` makes much more sense TBH...

theseer replied:
Wouldn't that be `${PLACEHOLDER}`

sebastianfeldmann replied:
I changed the placeholder syntax to `{$PLACEHOLDER}` and set the xsd type to `xs:anyURI`

sebastianfeldmann replied:
Ahhh I have no preference, I could change it to `${PLACEHOLDER}` in no time.

theseer replied:
Probably depends on what syntax / language style we want to follow.

In a PHP string, it would be `$varname`, whith `{$varnbame}` being an alternative.
In BASH, ANT's `build.xml` and alike it's `${varname}`.

sebastianfeldmann replied:
Since it's a PHP lib I would argue to go with `{$varname}` but... like I said... no preference here :)

theseer replied:
`{{varname}}`

/me runs

sebastianfeldmann replied:
Fine for me as well :P 
5 4 3 2 ...

theseer replied:
I somehow don't like the `$` in the URL ;)
So, unless @heiglandreas totally disagrees, let's go for mustache-style `{{varname}}`

sebastianfeldmann replied:
1 done placeholder syntax is now `{{varname}}`

heiglandreas reviewed (dismissed):
We can modify this later at any time. From my side this can be merged.

sebastianfeldmann committed:
Change placeholder syntax

In order to allow xs:anyURI for the xsd type definition the
url placeholder had to be changed vom %placeholder% to
{$placeholder}.
